### ISSUE TITLE
Harsh noise when playing on speakers (Linux)

### ISSUE THREAD

**Issue by Fussmatte** [TerryCavanagh/VVVVVV#970]
This has affected me for a good while now. If I have the game open at all, then harsh noises play out over any audio that's currently playing. These noises don't occur during silence, but if I have the game on mute then if anything else plays audio the noises will happen. This only happens with VVVVVV specifically, and on Linux. These noises do not get picked up by screen recorders, so here's a phone recording I took a while back:

https://user-images.githubusercontent.com/44736084/233127653-a8a25d91-0753-484f-bcef-bd5a065977cb.mp4

An extra note: sometimes, the noises will stop altogether during a playthrough, though not before finishing off with an especially harsh noise (as if it was 'flushing out' the problem causing them).

**Comment by flibitijibibo** [comment]
One quick test is to run the game with `SDL_AUDIODRIVER=disk`, which will generate raw float32 PCM that something like Audacity (or whatever's left of Audacity...) can import as raw data. If that comes out clean, it's likely something related to Pulse, and at that point `SDL_AUDIODRIVER=pipewire` may produce a different result.

**Comment by TerryCavanagh** [comment]
Wow!

(Does this happen with v2.3, I wonder? I haven't seen any reports of this before, so it makes sense that it's due to a change since the last public release)

**Comment by flibitijibibo** [comment]
It's possible that the change from SDL_mixer to FAudio changed the app-side mixing, but since it's device-specific that would have to be within SDL, since this would mean that the final mix is being mangled between FAudio and the final output. SDL2's resampler is "eh", but not _that_ eh, so I'm wondering if Pulse doesn't like that FAudio's update size is smaller than SDL_mixer's (FAudio's is about 1/10th SDL_mixer's, for precision/accuracy reasons) and the laptop isn't happy about it. Usually this gets worked around by increasing the OS' latency a bit, but I wouldn't be surprised if using PipeWire was enough to fix the issue properly.

**Comment by Fussmatte** [comment]
> 

It doesn't happen in 2.3, no.

**Comment by Fussmatte** [comment]
The raw audio that `SDL_AUDIODRIVER=disk` gives doesn't have the noises in it. Lines up with what happens when I try to record it with OBS or whatever else. When I have the option off the noises still happen though. `SDL_AUDIODRIVER=pipewire` just complains that it can't open an audio device

**Comment by flibitijibibo** [comment]
> The raw audio that `SDL_AUDIODRIVER=disk` gives doesn't have the noises in it. Lines up with what happens when I try to record it with OBS or whatever else. When I have the option off the noises still happen though. `SDL_AUDIODRIVER=pipewire` just complains that it can't open an audio device

That pretty much settles it - Pulse is giving us sass! A couple options:

- Update the system to one that uses PipeWire by default; I endorse [Fedora](https://fedoraproject.org/) but pretty much everybody has moved to PW; Debian 12 is last on the list and will likely be out this summer.
- Try `PULSE_LATENCY_MSEC=60` (or higher, depends on how flaky the sound driver is)

**Comment by Fussmatte** [comment]
First one is not even an option for me, I'm not uprooting my entire OS setup just for a game. Second one seems to work fine, the noises don't occur anymore until I run it without the option

**Comment by NyakoFox** [comment]
I've been trying to fix audio playback on a Switchbrew port of VVVVVV (buffer issues basically; music is slower & buzzy), and for some reason, the same fix commented by @midzer also fixes buffer issues on the Nintendo Switch.

The part that confuses me is the original code; It basically does `samples / 100`. which ends up being something like `441` if the source file is 44.1 khz. Apparently, the original line did something like `want.samples = 1024;`, and it was changed in a commit. 1024 DOES work, but 441 does not... (I also tested 512, which doesn't work either.)

@flibitijibibo Maybe this would be a better question in the FAudio repository rather than here, but I'm curious behind this FAudio change? It seems to cause problems in both a WASM port and my Switchbrew port; I don't know much about audio, so I might be completely missing the mark here, but some help here would be very appreciated!

**Comment by flibitijibibo** [comment]
The buffer size is an XAudio2 thing - for whatever reason it expects updates to be mixed in 10ms chunks; not doing this causes some effects to crash since it assumes the 10ms window.

One possible workaround is to use FAUDIO_1024_QUANTUM, but for some odd reason XAudio chose to make this represent a 21.33ms window rather than just a 1024-sample window, so the sample rate would need to be set to 48KHz for it to be consistent.

**Comment by NyakoFox** [comment]
Do you think it'd be worth it to add a flag for a static 1024-sample window?

I understand the importance of accuracy for FNA; if this is a problem that exists outside of that, then maybe games which aren't using FNA but only FAudio would benefit from some sort of configuration like this?

(Also as a follow-up, `FAUDIO_1024_QUANTUM` does seem to fix the problem; even if using 44.1k, it seems fine as far as I'm aware. It seems like the old buffer was just too small, and ~940 is good enough.)

**Comment by flibitijibibo** [comment]
If adding that flag fixes the issue we can apply that!

**Comment by NyakoFox** [comment]
We would probably want to verify with someone who has the problem on Linux; though I think @Fussmatte has changed distros by now... though it probably wouldn't hurt to push it blindly since it most likely only has upsides?

**Comment by flibitijibibo** [comment]
Yeah as long as it works on at least one setup that should be enough - if there are still issues it'll likely be something other than the buffer size.
